[solc] colorized diagnostics output

christianparpart opened the pull request:
This PR is a resurrection of #3046 and their related #4338 and #4340.

* ANSI coloring (termcolor): I didn't copy'n'paste the original one (which requires a additional license), but hacked together our own, that fits our needs
* SourceReferenceFormatter: greatly refactored to:
  * have improved diagnostics output format (this includes also streamlining line-truncation, as mentioned in #5169 (which we can close then).
  * support colored output
* solc: adds `--color` for forced colorized output and `--no-color` for explicitly disabling terminal-autodetection, e.g. if none of these will be given, the diagnostic output will be colored when stdout/stderr is connected to a terminal.

![image](https://user-images.githubusercontent.com/56763/49053534-cdae5700-f1f0-11e8-9d2f-c7440d8a0203.png)

### please note

**Only the last 3 commits matter**, the rest is from the rebase.

Commit:
liblangutil: refactors SourceReferenceFormatter error formatting for pretty and colored output.

* Refactors output format in a way it is (or should at least be) more readable.
  (NB.: As source of inspiration, I chose the rustc compiler output.)
* Adds color support to the stream output.
* Also improves multiline source formatting
  (i.e. truncating too long lines, like done with single lines already)
* solc: adds flags --color (force terminal colors) and --no-color (disable autodetection)
* solc: adds --new-reporter to give output in *new* formatting (colored or not)
* Changelog adapted accordingly.

chriseth commented:
The test failure is not about strcmp: `Unknown exception during compilation.`. Also colouring should be disabled whenever output does not go to a terminal.

chriseth commented:
Also, could you split up source reference formatter into two components: One that just does source extraction, cropping and other things (and returns structured data) and one that adds the colour formatting?

christianparpart commented:
@chriseth it was late night (almost morning? :-D). I just noticed tests failing with some error. I'll look into that tomorrow. Thanks :)

WRT not enabling color when output streams aren't a terminal, that's actually already happening (unless there is a bug inside), you can force colors regardless of that with `--color` CLI option to solc, or force disabling (i.e. no auto-enabling if stdout/stderr is connected to terminal).

The source extraction split-out I can do.

chriseth commented:
Sounds good, but please split it up.

chriseth commented on test/cmdlineTests/too_long_line_edge_out.sol.err:
```diff
@@ -1,6 +1,13 @@
-too_long_line_edge_out.sol:2:37: Error: Identifier not found or not unique.
- ...   function fffffffffffffffffffffff(announcementTypeTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT Typ, string A) onlyOwner external  ...
-                                        ^----------------------------------------------------------------------------------------------^
+
```
Should not add an empty line at the beginning.

chriseth commented on test/cmdlineTests/too_long_line_edge_out.sol.err:
```diff
@@ -1,6 +1,13 @@
+
+Warning: Source file does not specify required compiler version!
+
+ --> too_long_line_edge_out.sol:1:1: Source file does not specify required compiler version!
```
Do we really need to repeat the message?

christianparpart replied:
no. this output just hasn't been updated to the code changes yet. will fix that. thanks :)

chriseth commented on test/cmdlineTests/too_long_line_edge_out.sol.err:
```diff
@@ -1,6 +1,13 @@
+Warning: Source file does not specify required compiler version!
+
+ --> too_long_line_edge_out.sol:1:1: Source file does not specify required compiler version!
+  |
```
Is this line always empty?

chriseth commented on test/cmdlineTests/too_long_line_edge_out.sol.err:
```diff
@@ -1,6 +1,13 @@
+
+ --> too_long_line_edge_out.sol:1:1: Source file does not specify required compiler version!
+  |
+0 | contract C {
```
Even though we count lines starting from 0 internally, I think it should say 1 here.

christianparpart replied:
that's a bug I fixed already too. need to reflect the changes in the test suite. :)

chriseth commented on libdevcore/termcolor.cpp:
```diff
@@ -0,0 +1,250 @@
+
+bool is_terminal(ostream const& os)
+{
+	if (&os == &cout)
```
This looks very weird to me.

chriseth commented on libdevcore/termcolor.cpp:
```diff
@@ -0,0 +1,250 @@
+
+bool is_enabled(ostream const& os)
+{
+	return const_cast<ostream&>(os).iword(internal::enabled_index) != 0;
```
This looks rather magical. Is there no better way to achieve the same?

christianparpart replied:
That's actually how the way to go (it seems) to plug custom flags to a stream.

https://en.cppreference.com/w/cpp/io/ios_base/iword

chriseth commented on libdevcore/termcolor.cpp:
```diff
@@ -0,0 +1,250 @@
+
+ostream& enable(ostream& os)
+{
+	if (!is_terminal(os))
```
Shouldn't this read `if (is_terminal` instead?

christianparpart replied:
No, because while stdout (cout) can be redirected by the users, it doesn't mean the user has redirected stderr (cerr), for example.

Just like the libc call `isatty(fd)` requires a file descriptor to check, too).

chriseth commented on libdevcore/termcolor.h:
```diff
@@ -0,0 +1,105 @@
+// background colors
+internal::CodeSeq background(uint8_t code);
+internal::CodeSeq background(color c);
+std::ostream& on_black(std::ostream& os);
```
Coding style asks for `onBlack`

christianparpart replied:
I absolutely agree, and I'd have done that myself in app code, but with termcolor API, it felt more natural to use the C++ standard library's naming convention. What do you think?

chriseth commented on liblangutil/SourceReferenceFormatterHuman.cpp:
```diff
@@ -0,0 +1,114 @@
+
+	// line 0: source name
+	m_stream << string(leftpad, ' ') << frameColor << "--> " << termcolor::reset
+			 << _ref.sourceName << ":" << (_ref.position.line + 1) << ":" << (_ref.position.column + 1) << ": "
```
space after tab

chriseth commented on liblangutil/SourceReferenceFormatterHuman.cpp:
```diff
@@ -0,0 +1,114 @@
+		int const locationLength = _ref.endColumn - _ref.startColumn;
+
+		// line 1:
+		m_stream << string(leftpad, ' ') << frameColor << " |" << termcolor::reset << '\n';
```
Why would you print an empty line? To have spacing and make the source snippet stand out?

christianparpart replied:
For the diagnostics output format I got literally inspired by the output format of the Rust compiler (just like the original PR from the external contribution).

Apart from that, I feel that more space around makes the actual text also more readable.

chriseth commented on liblangutil/SourceReferenceFormatterHuman.h:
```diff
@@ -0,0 +1,65 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author Christian <c@ethdev.com>
```
I guess this could be removed.

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -691,6 +723,12 @@ Allowed options)",
 
+	if (!setupTermColors(m_args.count(g_argColor), m_args.count(g_argNoColor)))
+	{
+		serr() << "Option " << g_argColor << " and " << g_argNoColor << " are mutualy exclusive." << endl;
```
Isn't there a boost program options feature for that?

christianparpart replied:
I could only find one for flags in general.

But what I need is some kind of an **optional** flag (like a tribool), so I can reflect (yes, no, auto-detect).

chriseth commented on libdevcore/termcolor.cpp:
```diff
@@ -0,0 +1,240 @@
+	// Set output mode to handle virtual terminal (ANSI escape sequences)
+	// ignore any error, as this is just a "nice-to-have"
+	// only windows needs to be taken care of, as other platforms (Linux/OSX) support them natively.
+	HANDLE hOut = GetStdHandle(fdMap[terminalFd]);
```
This assumes that `STDOUT_FILENO` is at most 2. You could use an `std::array` and `.at()`.

chriseth commented on Changelog.md:
```diff
@@ -14,6 +14,8 @@ Compiler Features:
  * Static Analyzer: Do not warn about unused variables or state mutability for functions with an empty body.
  * Type Checker: Add an additional reason to be displayed when type conversion fails.
  * Yul: Support object access via ``datasize``, ``dataoffset`` and ``datacopy`` in standalone assembly mode.
+ * Commandline interface: Adds new option ``--new-reporter`` for improved diagnostics formatting
```
Needs to be moved.

christianparpart replied:
I assume you mean it should be alphabetically ordered. changed that. (but other entries seem to conflict, too, I realized)

chriseth commented on liblangutil/SourceReferenceFormatterHuman.h:
```diff
@@ -0,0 +1,64 @@
+
+		SourceReferenceFormatterHuman formatter(errorOutput);
+		formatter.printExceptionInformation(_exception, _name);
+		return errorOutput.str();
```
This function uses ostringstream, so it might bypass the init functionality if it is based on streams. I think this is evidence that we should initialize the terminal once inside CommandlineInterface.cpp instead.

christianparpart replied:
The terminal initialization is now done before `main()` is being invoked so any use of streams is fine within main() and after (atexit handlers).

chriseth commented on liblangutil/SourceReferenceFormatterHuman.cpp:
```diff
@@ -0,0 +1,115 @@
+
+		// line 2:
+		m_stream
+			<< frameColor << (_ref.position.line + 1) << " | " << termcolor::reset
```
I think this file should really use the already existing `FormattedScope`. This would remove the need for in-stream-information-passing (whose concept looks a little strange to me in general) and would also auto-reset because of RAII.

chriseth commented on liblangutil/SourceReferenceFormatterHuman.cpp:
```diff
@@ -0,0 +1,115 @@
+
+	printSourceLocation(_msg.primary);
+
+	for (auto const& secondary : _msg.secondary)
```
No spaces in front of `:`

chriseth commented:
Is it possible to test the colour output (including the options to enable and disable it)?

chriseth commented on Changelog.md:
```diff
@@ -24,6 +24,7 @@ Compiler Features:
  * Code Generator: Only check callvalue once if all functions are non-payable.
  * Code Generator: Use codecopy for string constants more aggressively.
  * Code Generator: Use binary search for dispatch function if more efficient. The size/speed tradeoff can be tuned using ``--optimize-runs``.
+ * Commandline interface: Adds new option ``--new-reporter`` for improved diagnostics formatting along with ``--color`` and ``--no-color`` for colorized output to be forced (or explicitly disabled).
```
Should move to 0.5.3

christianparpart replied:
agreed.

chriseth commented:
I really think this should use FormattedScope instead. We can perhaps discuss next week.

Commit:
libdevcore: Introduces a generalized AnsiColorized, an improved FormattedScope

a future commit/PR could replace existing code to use AnsiColorized and
remove the old implementation of FormattedScope.

chriseth commented on libdevcore/AnsiColorized.h:
```diff
@@ -0,0 +1,119 @@
+
+	AnsiColorized& operator=(AnsiColorized&& other)
+	{
+		m_stream = other.m_stream;
```
Please add a comment about how this is different from the default move assignment.

chriseth commented on liblangutil/SourceReferenceFormatterHuman.h:
```diff
@@ -0,0 +1,80 @@
+	}
+
+private:
+	inline dev::AnsiColorized normalColor() const { return dev::AnsiColorized(m_colored, {dev::formatting::WHITE}); };
```
Why inline?

chriseth commented on liblangutil/SourceReferenceFormatterHuman.cpp:
```diff
@@ -0,0 +1,106 @@
+		for_each(
+			_ref.text.cbegin(),
+			_ref.text.cbegin() + _ref.startColumn,
+			[this](char ch) { m_stream << (ch == '\t' ? '\t' : ' '); }
```
Nice! We might even have a bug about this.

christianparpart replied:
I too actually held back for a moment on that line, but said _if_ that's a bug, then it was there before already, and I'd prefer to address it in a separate PR (as the base class is affected then, too).

ekpyron commented on libdevcore/AnsiColorized.h:
```diff
@@ -0,0 +1,91 @@
+}
+
+/// AnsiColorized provides a convenience helper to colorize ostream with formatting-reset assured.
+class AnsiColorized
```
Should we delete the copy constructor and default the move constructor for this :-)? (or even implement a move that sets ``m_enabled`` to false in the argument?) It won't be a real problem, but if this is copied, I guess there will be unnecessary output to the stream... currently there shouldn't be any copies, so this is a hypothetic non-issue, but maybe we should disallow that anyways.
But regard this comment as entirely "optional" :-).

ekpyron reviewed (dismissed):
Looks good now!

chriseth commented:
I think there is still an inconsistency with regards to spacing between the errors.
This is how errors with secondary source locations are formatted:
```
Warning: Source file does not specify required compiler version! Consider adding "pragma solidity ^0.5.4;"
 --> /tmp/x.sol:1:1: 
  |
1 | contract c {
  | ^ (Relevant source part starts here and spans across multiple lines).
Error: Identifier already declared.
 --> /tmp/x.sol:3:1: 
  |
3 | int x = 3;}
  | ^^^^^

Note: The previous declaration is here:
 --> /tmp/x.sol:2:32: 
  |
2 | function f(uint) public pure { int x = 2;
  |                                ^^^^^

Error: No matching declaration found after argument-dependent lookup.
 --> /tmp/x.sol:5:27: 
  |
5 | function g() public pure{ f();}
  |                           ^

```

And this is how it looks without an error with secondary source location:
```
Warning: Source file does not specify required compiler version! Consider adding "pragma solidity ^0.5.4;"
 --> /tmp/x.sol:1:1: 
  |
1 | contract c {
  | ^ (Relevant source part starts here and spans across multiple lines).
Error: No matching declaration found after argument-dependent lookup.
 --> /tmp/x.sol:4:27: 
  |
4 | function g() public pure{ f();}
  |                           ^

```

I would have expected the secondary source location to be appended without an empty line between the main location and the secondary location.

Also we might want to add an empty line between two errors (actually this is how your screenshot in the description has it).

ekpyron commented:
If we make more changes to this anyways, then I vote for changing the command line argument from ``--new-reporter`` to ``--fancy-reporter`` as @chriseth just suggested in the office ;-).

chriseth commented:
Oh and another thing I just noticed: At least in the screenshot in the description, the line numbering it says `t.sol:3:...` but then on the left there is a `2`. I think lines should be numbered starting from 1 in both cases.

chriseth commented:
Ah ok, but the thing I pasted at least has it right.

chriseth commented:
Output is now as follows:
```
> solc --new-reporter /tmp/x.sol 
Warning: This is a pre-release compiler version, please do not use it in production.

Error: No visibility specified. Did you intend to add "public"?
 --> /tmp/x.sol:2:1: 
  |
2 | function f() {}
  | ^^^^^^^^^^^^^^^

Error: No visibility specified. Did you intend to add "public"?
 --> /tmp/x.sol:6:1: 
  |
6 | function h() {}
  | ^^^^^^^^^^^^^^^

Warning: Source file does not specify required compiler version!
 --> /tmp/x.sol:1:1: 
  |
1 | contract c {
  | ^ (Relevant source part starts here and spans across multiple lines).

Error: Function with same name and arguments defined twice.
 --> /tmp/x.sol:2:1: 
  |
2 | function f() {}
  | ^^^^^^^^^^^^^^^
Note: Other declaration is here:
 --> /tmp/x.sol:5:1: 
  |
5 | function f() public pure {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Overriding function changes state mutability from "pure" to "nonpayable".
 --> /tmp/x.sol:2:1: 
  |
2 | function f() {}
  | ^^^^^^^^^^^^^^^
Note: Overridden function is here:
 --> /tmp/x.sol:5:1: 
  |
5 | function f() public pure {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

ekpyron reviewed (dismissed):
Looks good!

chriseth commented:
I also verified that the old reporter is used with standard-json.

chriseth commented:
Rebased.

christianparpart commented:
P.S. @chriseth I didn't rename it to `--fancy-reporter`, even though I really very much liked the idea, but I'd rather prefer to see this reporter becoming standard at some point in the future (not necessarily the very next breaking release)

chriseth commented:
I guess the plan is to make it the default starting from 0.6.0
